crates/time/src/lib.rs

```diff
@@ -0,0 +1,319 @@
+                return Ok(slot)
+            }
+        }
+        return Err(TimeHorizonError::PastTimeHorizon)
+    }
+
+    pub fn slot_to_epoch(&self, slot: u64) -> Result<u64, TimeHorizonError> {
+        for era in &self.eras {
+            if era.start.slot > slot {
+                return Err(TimeHorizonError::InvalidEraHistory)
+            }
+            if era.end.slot >= slot {
+                let slots_elapsed = slot - era.start.slot;
+                let epochs_elapsed = slots_elapsed / era.params.epoch_size;
+                let epoch_number = era.start.epoch + epochs_elapsed;
+                return Ok(epoch_number)
+            }
+        }
+        return Err(TimeHorizonError::PastTimeHorizon)
+    }
+
+    pub fn epoch_bounds(&self, epoch: u64) -> Result<EpochBounds, TimeHorizonError> {
+        for era in &self.eras {
+            if era.start.epoch > epoch {
+                return Err(TimeHorizonError::InvalidEraHistory)
+            }
+            // We can't answer queries about the upper bound epoch of the era because the bound is
+            // exclusive.
+            if era.end.epoch > epoch {
+                let epochs_elapsed = epoch - era.start.epoch;
+                let offset = era.start.slot;
+                let start = offset + era.params.epoch_size * epochs_elapsed;
+                let end = offset + era.params.epoch_size * (epochs_elapsed + 1);
+                return Ok(EpochBounds {
+                    start: start,
+                    end: end,
+                })
+            }
+        }
+        return Err(TimeHorizonError::PastTimeHorizon);
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use hex;
+
+    #[test]
+    fn test_slot_to_time() {
+        let params = EraParams::new(86400, 1000).unwrap();
+        let eras = EraHistory {
+            eras: vec![
+                Summary {
+                    start: Bound {
+                        time: 0,
+                        slot: 0,
+                        epoch: 0,
+                    },
+                    end: Bound {
+                        time: 86400000,
+                        slot: 86400,
+                        epoch: 1,
+                    },
+                    params: params.clone(),
+                },
+                Summary {
+                    start: Bound {
+                        time: 86400000,
+                        slot: 86400,
+                        epoch: 1,
+                    },
+                    end: Bound {
+                        time: 172800000,
+                        slot: 172800,
+                        epoch: 2,
+                    },
+                    params: params.clone(),
+                },
+            ],
+        };
+        let t0 = eras.slot_to_relative_time(172801);
+        assert_eq!(t0, Err(TimeHorizonError::PastTimeHorizon));
+        let t1 = eras.slot_to_relative_time(172800);
+        assert_eq!(t1, Ok(172800000));
```

Comment:
I am not a big fan of testing several things in a single test: When the test fails, it's hard to know quickly what's failed, you would have to look at the source code of the test to know exactly what broke, which is annoying.

I would suggest to:
* not prefix test functions with `test_`,  we know they are tests because there's an annotation just above
* have more explicit test names (eg. `slot_to_time_fails_given_its_past_upper_bound`)
* have each function test a single "thing" (possibly using parameterised tests to avoid duplication)